`sepcomp-lib-lto` test fails on master with --enable-debug

When compiling with --enable-debug, the `sepcomp-lib-lto` test fails. It passes if I build without --enable-debug. Relevant build output:

```
$ git rev-parse HEAD
a9515698fa456390386087ccb6123ce741f18527
$ uname -a
Linux dev 3.19.0-18-generic #18-Ubuntu SMP Tue May 19 18:31:35 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
$ ./configure --enable-debug
$ make -j12 check-stage1-rpass TESTNAME=sepcomp-lib-lto

...

running 1 test
test [run-pass] run-pass/sepcomp-lib-lto.rs ... FAILED

failures:

---- [run-pass] run-pass/sepcomp-lib-lto.rs stdout ----

error: compilation failed!
status: signal: 6
command: x86_64-unknown-linux-gnu/stage1/bin/rustc /home/ubuntu/rust/src/test/run-pass/sepcomp-lib-lto.rs -L x86_64-unknown-linux-gnu/test/run-pass/ --target=x86_64-unknown-linux-gnu -L x86_64-unknown-linux-gnu/test/run-pass/sepcomp-lib-lto.stage1-x86_64-unknown-linux-gnu.run-pass.libaux -o x86_64-unknown-linux-gnu/test/run-pass/sepcomp-lib-lto.stage1-x86_64-unknown-linux-gnu -O -L x86_64-unknown-linux-gnu/rt -C lto
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
rustc: /home/ubuntu/rust/src/llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp:708: llvm::DIE* llvm::DwarfUnit::getOrCreateTypeDIE(const llvm::MDNode*): Assertion `Ty == resolve(Ty->getRef()) && "type was not uniqued, possible ODR violation."' failed.

------------------------------------------

thread '[run-pass] run-pass/sepcomp-lib-lto.rs' panicked at 'explicit panic', /home/ubuntu/rust/src/compiletest/runtest.rs:1497



failures:
    [run-pass] run-pass/sepcomp-lib-lto.rs

    test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured

    thread '<main>' panicked at 'Some tests failed', /home/ubuntu/rust/src/compiletest/compiletest.rs:252
    /home/ubuntu/rust/mk/tests.mk:756: recipe for target 'tmp/check-stage1-T-x86_64-unknown-linux-gnu-H-x86_64-unknown-linux-gnu-rpass.ok' failed
    make: *** [tmp/check-stage1-T-x86_64-unknown-linux-gnu-H-x86_64-unknown-linux-gnu-rpass.ok] Error 101
```


Backtrace:

```
#0  0x00007ffff6cb0267 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55  
#1  0x00007ffff6cb1eca in __GI_abort () at abort.c:89  
#2  0x00007ffff6ca903d in __assert_fail_base (fmt=0x7ffff6e0b028 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",   
    assertion=assertion@entry=0x7ffff1e75ca8 "Ty == resolve(Ty->getRef()) && \"type was not uniqued, possible ODR violation.\"", file=file@entry=0x7ffff1e75878 "/home/ubuntu/rust/src/llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp",   
    line=line@entry=708,   
    function=function@entry=0x7ffff1e78300 <llvm::DwarfUnit::getOrCreateTypeDIE(llvm::MDNode const*)::__PRETTY_FUNCTION__> "llvm::DIE* llvm::DwarfUnit::getOrCreateTypeDIE(const llvm::MDNode*)") at assert.c:92  
#3  0x00007ffff6ca90f2 in __GI___assert_fail (  
    assertion=0x7ffff1e75ca8 "Ty == resolve(Ty->getRef()) && \"type was not uniqued, possible ODR violation.\"",   
    file=0x7ffff1e75878 "/home/ubuntu/rust/src/llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp", line=708,   
    function=0x7ffff1e78300 <llvm::DwarfUnit::getOrCreateTypeDIE(llvm::MDNode const*)::__PRETTY_FUNCTION__> "llvm::DIE* llvm::DwarfUnit::getOrCreateTypeDIE(const llvm::MDNode*)") at assert.c:101  
#4  0x00007ffff0dcdd6e in llvm::DwarfUnit::getOrCreateTypeDIE(llvm::MDNode const*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#5  0x00007ffff0dcddb9 in llvm::DwarfUnit::addType(llvm::DIE&, llvm::DIType const*, llvm::dwarf::Attribute) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#6  0x00007ffff0dce59a in llvm::DwarfUnit::constructTypeDIE(llvm::DIE&, llvm::DISubroutineType const*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#7  0x00007ffff0dcdc7a in llvm::DwarfUnit::getOrCreateTypeDIE(llvm::MDNode const*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#8  0x00007ffff0dcddb9 in llvm::DwarfUnit::addType(llvm::DIE&, llvm::DIType const*, llvm::dwarf::Attribute) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#9  0x00007ffff0dcd3af in llvm::DwarfUnit::applySubprogramAttributes(llvm::DISubprogram const*, llvm::DIE&, bool) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#10 0x00007ffff0da7962 in llvm::DwarfCompileUnit::applySubprogramAttributesToDefinition(llvm::DISubprogram const*, llvm::DIE&) () from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#11 0x00007ffff0daebd2 in llvm::DwarfCompileUnit::constructAbstractSubprogramScopeDIE(llvm::LexicalScope*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#12 0x00007ffff0dbc4f5 in llvm::DwarfDebug::constructAbstractSubprogramScopeDIE(llvm::LexicalScope*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#13 0x00007ffff0dc210b in llvm::DwarfDebug::endFunction(llvm::MachineFunction const*) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#14 0x00007ffff0d8dbe5 in llvm::AsmPrinter::EmitFunctionBody() ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#15 0x00007ffff0a925c6 in llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#16 0x00007ffff16b71a8 in llvm::FPPassManager::runOnFunction(llvm::Function&) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#17 0x00007ffff16b7893 in llvm::legacy::PassManagerImpl::run(llvm::Module&) ()  
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_llvm-d8ace771.so  
#18 0x00007ffff047d095 in LLVMRustWriteOutputFile (Target=0x7fffe8085580, PMR=0x7fffe80211f0, M=0x7fffe8006220,   
    path=0x7fffed3f4b20 "x\326\034\356\377\177", FileType=llvm::TargetMachine::CGFT_ObjectFile)  
    at /home/ubuntu/rust/src/rustllvm/PassWrapper.cpp:249  
#19 0x00007ffff55de42a in rustc_trans::back::write::write_output_file (handler=0x7fffed3f84c8, target=0x7fffe8085580,   
    pm=0x7fffe80211f0, m=0x7fffe8006220, output=0x7fffec4e2d80, file_type=ObjectFileType)  
    at src/librustc_trans/back/write.rs:72  
#20 0x00007ffff55eb33f in fnfn (cpm=0x7fffe80211f0) at src/librustc_trans/back/write.rs:565  
#21 0x00007ffff55eb26e in rustc_trans::back::write::optimize_and_codegen::with_codegen<closure> (tm=0x7fffe8085580,   
    llmod=0x7fffe8006220, no_builtins=false, f={void (struct (*mut rustc_llvm::PassManager_opaque))} 0x7fffed3f4e68)  
    at src/librustc_trans/back/write.rs:533  
#22 0x00007ffff55ea9cb in fnfn () at src/librustc_trans/back/write.rs:564  
#23 0x00007ffff55ea0ed in rustc_trans::util::common::time<(),(),closure> (do_it=false, what=..., u=0,   
    f={void (struct (()))} 0x7fffed3f5460) at src/librustc/util/common.rs:39  
#24 0x00007ffff55e80ea in rustc_trans::back::write::optimize_and_codegen (cgcx=0x7fffed3f63c0, mtrans=..., config=...,   
    name_extra=..., output_names=...) at src/librustc_trans/back/write.rs:543  
#25 0x00007ffff55f2864 in rustc_trans::back::write::execute_work_item (cgcx=0x7fffed3f63c0, work_item=...)  
    at src/librustc_trans/back/write.rs:906  
#26 0x00007ffff55ef155 in rustc_trans::back::write::run_work_singlethreaded (sess=0x7fffed3f7c88, reachable=...,   
    work_items=...) at src/librustc_trans/back/write.rs:919  
#27 0x00007ffff55ebe6f in rustc_trans::back::write::run_passes (sess=0x7fffed3f7c88, trans=0x7fffed3f7968,   
    output_types=..., crate_output=0x7fffed3f7a10) at src/librustc_trans/back/write.rs:677
#28 0x00007ffff7a6c7fe in fnfn () at src/librustc_driver/driver.rs:767
#29 0x00007ffff7a6c432 in rustc_driver::util::common::time<(),(),closure> (do_it=false, what=..., u=0, 
    f={void (struct (()))} 0x7fffed3f74e8) at src/librustc/util/common.rs:39
#30 0x00007ffff78ed1fd in rustc_driver::driver::phase_5_run_llvm_passes (sess=0x7fffed3f7c88, trans=0x7fffed3f7968, 
    outputs=0x7fffed3f7a10) at src/librustc_driver/driver.rs:766
#31 0x00007ffff7844144 in rustc_driver::driver::compile_input (sess=..., cfg=..., input=0x7fffed3fdf98, 
    outdir=0x7fffed3fe110, output=0x7fffed3fe0f0, addl_plugins=..., control=...) at src/librustc_driver/driver.rs:168
#32 0x00007ffff7b096df in rustc_driver::run_compiler (args=..., callbacks=...) at src/librustc_driver/lib.rs:156
#33 0x00007ffff7b060fc in fnfn () at src/librustc_driver/lib.rs:99
#34 0x00007ffff7b052c9 in fnfn () at src/librustc_driver/lib.rs:806
#35 0x00007ffff7b051a6 in rustc_driver::boxed::F.FnBox<A>::call_box (self=0x7fffed43a060, args=0)
    at src/liballoc/boxed.rs:398
#36 0x00007ffff7b04a6f in rustc_driver::boxed::Box<FnBox<A, Output = R>+ Send + 'a>.FnOnce<A>::call_once (self=..., 
    args=0) at src/liballoc/boxed.rs:414
#37 0x00007ffff7b044cc in fnfn () at src/libstd/thread/mod.rs:349
#38 0x00007ffff7b0444f in rustc_driver::rt::unwind::try::try_fn<closure> (opt_closure=0x7fffed3fe8e8)
    at src/libstd/rt/unwind/mod.rs:158
#39 0x00007ffff7b043ea in rt::unwind::try::try_fn::h17428172142314623289 ()
   from x86_64-unknown-linux-gnu/stage1/lib/librustc_driver-d8ace771.so
#40 0x00007ffff7332de9 in rust_try_inner () from x86_64-unknown-linux-gnu/stage1/lib/libstd-d8ace771.so
#41 0x00007ffff7332dd6 in rust_try () from x86_64-unknown-linux-gnu/stage1/lib/libstd-d8ace771.so
#42 0x00007ffff718fa30 in std::rt::unwind::try::inner_try (f={void (enum class c_void *)} 0x7fffed3fe8a8, 
    data=0x7fffed3fe8e8) at src/libstd/rt/unwind/mod.rs:147
#43 0x00007ffff7b04385 in rustc_driver::rt::unwind::try<closure> (f={void (())} 0x7fffed3fe900)
    at src/libstd/rt/unwind/mod.rs:130
#44 0x00007ffff7b0419d in fnfn () at src/libstd/thread/mod.rs:349
#45 0x00007ffff7b04ca1 in rustc_driver::boxed::F.FnBox<A>::call_box (self=0x7fffed424180, args=0)
    at src/liballoc/boxed.rs:398
#46 0x00007ffff71b0eaf in std::boxed::Box<FnBox<A, Output = R>+ 'a>.FnOnce<A>::call_once (self=..., args=0)
    at src/liballoc/boxed.rs:406
#47 0x00007ffff71b0de7 in std::sys_common::thread::start_thread (main=0x7fffed43b000)
    at src/libstd/sys/common/thread.rs:30
#48 0x00007ffff71e2d05 in std::sys::thread::Thread::new::thread_start (main=0x7fffed43b000)
    at src/libstd/sys/unix/thread.rs:77
#49 0x00007ffff71e2c85 in sys::thread::Thread::new::thread_start::hfcb7277d7665b775tKv ()
   from x86_64-unknown-linux-gnu/stage1/lib/libstd-d8ace771.so
#50 0x00007fffef6db6aa in start_thread (arg=0x7fffed3ff700) at pthread_create.c:333
#51 0x00007ffff6d81eed in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

--- a/src/librustc_trans/trans/debuginfo/metadata.rs
+++ b/src/librustc_trans/trans/debuginfo/metadata.rs
@@ -63,8 +63,8 @@ pub const UNKNOWN_LINE_NUMBER: c_uint = 0;
 pub const UNKNOWN_COLUMN_NUMBER: c_uint = 0;
 
 // ptr::null() doesn't work :(
-const UNKNOWN_FILE_METADATA: DIFile = (0 as DIFile);
-const UNKNOWN_SCOPE_METADATA: DIScope = (0 as DIScope);
+const NO_FILE_METADATA: DIFile = (0 as DIFile);
+const NO_SCOPE_METADATA: DIScope = (0 as DIScope);
 
 const FLAGS_NONE: c_uint = 0;
 
@@ -566,7 +566,7 @@ fn vec_slice_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                                            &slice_type_name[..],
                                            unique_type_id,
                                            &member_descriptions,
-                                           UNKNOWN_SCOPE_METADATA,
+                                           NO_SCOPE_METADATA,
                                            file_metadata,
                                            span);
     return MetadataCreationResult::new(metadata, false);
@@ -611,7 +611,7 @@ fn subroutine_type_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
         unsafe {
             llvm::LLVMDIBuilderCreateSubroutineType(
                 DIB(cx),
-                UNKNOWN_FILE_METADATA,
+                NO_FILE_METADATA,
                 create_DIArray(DIB(cx), &signature_metadata[..]))
         },
         false);
@@ -655,7 +655,7 @@ fn trait_pointer_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                             unique_type_id,
                             &[],
                             containing_scope,
-                            UNKNOWN_FILE_METADATA,
+                            NO_FILE_METADATA,
                             codemap::DUMMY_SP)
 }
 
@@ -851,13 +851,6 @@ pub fn type_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
 }
 
 pub fn file_metadata(cx: &CrateContext, full_path: &str) -> DIFile {
-    match debug_context(cx).created_files.borrow().get(full_path) {
-        Some(file_metadata) => return *file_metadata,
-        None => ()
-    }
-
-    debug!("file_metadata: {}", full_path);
-
     // FIXME (#9639): This needs to handle non-utf8 paths
     let work_dir = cx.sess().working_dir.to_str().unwrap();
     let file_name =
@@ -867,6 +860,24 @@ pub fn file_metadata(cx: &CrateContext, full_path: &str) -> DIFile {
             full_path
         };
 
+    file_metadata_(cx, full_path, file_name, &work_dir)
+}
+
+pub fn unknown_file_metadata(cx: &CrateContext) -> DIFile {
+    // Regular filenames should not be empty, so we abuse an empty name as the
+    // key for the special unknown file metadata
+    file_metadata_(cx, "", "<unknown>", "")
+
+}
+
+fn file_metadata_(cx: &CrateContext, key: &str, file_name: &str, work_dir: &str) -> DIFile {
+    match debug_context(cx).created_files.borrow().get(key) {
+        Some(file_metadata) => return *file_metadata,
+        None => ()
+    }
+
+    debug!("file_metadata: file_name: {}, work_dir: {}", file_name, work_dir);
+
     let file_name = CString::new(file_name).unwrap();
     let work_dir = CString::new(work_dir).unwrap();
     let file_metadata = unsafe {
@@ -875,8 +886,8 @@ pub fn file_metadata(cx: &CrateContext, full_path: &str) -> DIFile {
     };
 
     let mut created_files = debug_context(cx).created_files.borrow_mut();
-    created_files.insert(full_path.to_string(), file_metadata);
-    return file_metadata;
+    created_files.insert(key.to_string(), file_metadata);
+    file_metadata
 }
 
 /// Finds the scope metadata node for the given AST node.
@@ -1227,7 +1238,7 @@ fn prepare_tuple_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                            tuple_llvm_type,
                            &tuple_name[..],
                            unique_type_id,
-                           UNKNOWN_SCOPE_METADATA),
+                           NO_SCOPE_METADATA),
         tuple_llvm_type,
         TupleMDF(TupleMemberDescriptionFactory {
             component_types: component_types.to_vec(),
@@ -1570,9 +1581,14 @@ fn prepare_enum_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                                    -> RecursiveTypeDescription<'tcx> {
     let enum_name = compute_debuginfo_type_name(cx, enum_type, false);
 
-    let (containing_scope, definition_span) = get_namespace_and_span_for_item(cx, enum_def_id);
-    let loc = span_start(cx, definition_span);
-    let file_metadata = file_metadata(cx, &loc.file.name);
+    let (containing_scope, _) = get_namespace_and_span_for_item(cx, enum_def_id);
+    // FIXME: This should emit actual file metadata for the enum, but we
+    // currently can't get the necessary information when it comes to types
+    // imported from other crates. Formerly we violated the ODR when performing
+    // LTO because we emitted debuginfo for the same type with varying file
+    // metadata, so as a workaround we pretend that the type comes from
+    // <unknown>
+    let file_metadata = unknown_file_metadata(cx);
 
     let variants = &enum_type.ty_adt_def().unwrap().variants;
 
@@ -1613,7 +1629,7 @@ fn prepare_enum_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                         DIB(cx),
                         containing_scope,
                         name.as_ptr(),
-                        UNKNOWN_FILE_METADATA,
+                        NO_FILE_METADATA,
                         UNKNOWN_LINE_NUMBER,
                         bytes_to_bits(discriminant_size),
                         bytes_to_bits(discriminant_align),
@@ -1765,7 +1781,7 @@ fn set_members_of_composite_type(cx: &CrateContext,
                     DIB(cx),
                     composite_type_metadata,
                     member_name.as_ptr(),
-                    UNKNOWN_FILE_METADATA,
+                    NO_FILE_METADATA,
                     UNKNOWN_LINE_NUMBER,
                     bytes_to_bits(member_size),
                     bytes_to_bits(member_align),
@@ -1808,7 +1824,7 @@ fn create_struct_stub(cx: &CrateContext,
             DIB(cx),
             containing_scope,
             name.as_ptr(),
-            UNKNOWN_FILE_METADATA,
+            NO_FILE_METADATA,
             UNKNOWN_LINE_NUMBER,
             bytes_to_bits(struct_size),
             bytes_to_bits(struct_align),
@@ -1869,7 +1885,7 @@ pub fn create_global_var_metadata(cx: &CrateContext,
         let loc = span_start(cx, span);
         (file_metadata(cx, &loc.file.name), loc.line as c_uint)
     } else {
-        (UNKNOWN_FILE_METADATA, UNKNOWN_LINE_NUMBER)
+        (NO_FILE_METADATA, UNKNOWN_LINE_NUMBER)
     };
 
     let is_local_to_unit = is_node_local_to_unit(cx, node_id);

